lib/test_ids/allocator.rb

```diff
@@ -14,6 +14,72 @@ def initialize(configuration)
+      # Not really using the Item.new currently, but instantiating the BinArray regardless
+      # for future enhancement/use.
+      r ||= TestIds::Configuration::Item.new
+      # Populate the BinArray with the range being passed.
+      r.include << range
+      # Check the database to see if the passed in range has already been included in the database hash
+      if rangehash.key?(:"#{range}")
+        # If the range already exists in the database, read out the last softbin given out for the flow across all ranges.
+        last_softbin = store['pointers']['softbin']
+        # Check if the last softbin given out is within the range that is being passed now.
+        if last_softbin.between?(range.min, range.max)
+          # If the last softbin is in range, calculate the pointer
+          # The pointer is calculated as a subtraction of the last_softbin from the range minimum value
+          # Ex: range = 10100..10199
+          #     last_softbin = 10101
+          #     @pointer = last_softbin - range.min (10101 - 10100) = 1
+          @pointer = last_softbin - range.min
+        else
+          # If the last_softbin is not in the given range, read out the database hash to see what
+          # the last_softbin given out was for that range.
+          # This hash is updated whenever a new softbin is assigned, so it should have the updated values for each range.
+          last_softbin = rangehash[:"#{range}"].to_i
+          # Now calculate the new pointer.
+          @pointer = last_softbin - range.min
+        end
+        # Check if the last_softbin given out is the same as the range[@pointer],
+        # if so increment pointer by softbin size, default size is 1, config.softbins.size is configurable.
+        # from example above, pointer was calculated as 1,range[1] is 10101 and is same as last_softbin, so pointer is incremented
+        # and new value is assigned to the softbin.
+        if last_softbin == range.to_a[@pointer]
+          @pointer += options[:size]
+          assigned_softbin = range.to_a[@pointer]
+        else
+          # Because of the pointer calculations above, I don't think it will ever reach here, has not in my test cases so far!
+          assigned_softbin = range.to_a[@pointer]
+        end
+        # Now update the database pointers to point to the lastest assigned softbin for a given range.
+        rangehash.merge!("#{range}": "#{range.to_a[@pointer]}")
+        # Also update the database so that the latest given softbin across all ranges is the latest.
+        store['pointers']['softbin'] = assigned_softbin
+      else
+        # This is the case for a brand new range that has not been passed before
+        # We start from the first value as the assigned softbin and update the database to reflect.
+        @pointer = 0
+        rangehash.merge!("#{range}": "#{range.to_a[@pointer]}")
+        assigned_softbin = range.to_a[@pointer]
+        store['pointers']['softbin'] = assigned_softbin
```

Comment:
Should not be updating this pointer, this is the pointer that is used by TestIDs whenever softbin is configured like:

~~~ruby
config.softbins.include << (1000..2000)
~~~

You should be creating a separate pointer here that lives under `store['pointers']['ranges'][<range_identifier>]`, possibly also with a key if you want to support the same range being used independently for two different purposes (e.g. bin and softbin) - `store['pointers']['ranges'][<key>][<range_identifier>]`